python/lsst/ts/ess/csc/data_client/young_32400_weather_station_data_client.py

```diff
@@ -0,0 +1,713 @@
+#
+# * wind_speed: raw counts; scale is sensor-specific
+# * wind_direction: relative angle in 10ths of a degree
+# * temperature: 0-4000 = 0-1V
+# * humidity: 0-4000 = 0-1V = 0-100% humidity
+# * pressure: 0-4000 = 0-1V; scale and offset are configurable in the sensor
+# * rain: 0-9999 (=MAX_RAIN_TIP_COUNT) tipping bucket tip count
+DATA_REGEX = re.compile(
+    r"(. )?(?P<wind_speed>\d\d\d\d) "
+    r"(?P<wind_direction>\d\d\d\d) "
+    r"(?P<temperature>\d\d\d\d) "
+    r"(?P<humidity>\d\d\d\d) "
+    r"(?P<pressure>\d\d\d\d) "
+    r"(?P<rain_tip_count>\d\d\d\d)"
+)
+
+
+class ScalingAccumulator:
+    """Accumulate values for a particular measurement.
+
+    Parameters
+    ----------
+    num_samples : `int`
+        Number of samples to accumulate. Must be positive.
+    scale : `float`
+        Scale to apply to raw values: scaled = raw * scale + offset
+    offset : `float`:
+        Offset to apply to raw values: scaled = raw * scale + offset
+    """
+
+    def __init__(self, num_samples: int, scale: float, offset: float) -> None:
+        if num_samples < 1:
+            raise ValueError(f"{num_samples=} must be >= 1")
+        self.num_samples = num_samples
+        self.scale = scale
+        self.offset = offset
+        self.values: list[float] = []
+
+    def add_sample(self, value: float) -> float | None:
+        """Add a raw value. Return the scaled median, if enough samples
+        have been accumulated, else None.
+
+        Parameters
+        ----------
+        raw : `float`
+            Raw value to scale and to accumulate.
+
+        Returns
+        -------
+        median : `float` | `None`
+            The median scaled value, if enough samples have been accumulated
+            (in which case the accumulator is reset) else None.
+        """
+        self.values.append(value)
+        if len(self.values) >= self.num_samples:
+            median = np.median(self.values)
+            self.values = []
+            return median * self.scale + self.offset
+        return None
+
+
+def float_to_intstr(value: float, max_int: int) -> str:
+    """Return a float value converted to an string representation
+    of an integer with 4 chars and leading zeros.
+
+    Parameters
+    ----------
+    value : `float`
+        The value to convert.
+    max_int : `int`
+        The maximum integer value. Must be > 0 and <= 9999.
+
+    Raises
+    ------
+    ValueError
+        If max_int <= 0 or > 9999.
+    """
+    if not 0 < max_int <= 9999:
+        raise ValueError(f"{max_int=} not >0 and <= 9999")
+    int_val = int(round(value))
+    truncated_int_val = max(0, min(int_val, max_int))
+    return f"{truncated_int_val:04d}"
+
+
+def scaled_from_raw(raw_value: float, scale: float, offset: float) -> float:
+    """Convert raw data to scaled: return raw * scale + offset."""
+    return raw_value * scale + offset
+
+
+class Young32400WeatherStationDataClient(common.BaseDataClient):
```

Comment:
I think this validates adding this class to ts_ess_common instead of here. I tried to keep the CSC functionality and the sensor/device definitions separate.

Reply:
I am uncomfortable with this. BaseDataClient is in ts_ess_common because it is used by multiple other packages.

The weather station data client is not common to anything; it is only used by the CSC.

I could see putting the weather station code in a new ts_ess_weatherstation[client] or ts_ess_dataclients package, but would be *very* reluctant to put it in package intended for common base code. The problem with ts_ess_weatherstation is that it's a lot of packaging for not very much code. ts_ess_dataclients makes sense, but what else would we want to move there, and on what timescale?